### ISSUE TITLE
Consider saving the code for labs

### ISSUE THREAD

**Issue by jdwilkin4** [freeCodeCamp/freeCodeCamp#57818]
A camper recently posted in the forum about their solutions for labs not being saved.

https://forum.freecodecamp.org/t/code-disappeared-from-full-stack-curriculum/727188

I do this this is something that should be discussed and considered on saving the solutions for labs so campers can go back.

Or

Maybe there should be a warning message before the camper goes to the next challenge.
The current modal does ask you if you want to download your solution.
But if the solutions aren't going to be saved, then we should consider a warning that if they don't save their solution it will be lost.

**Comment by a2937** [comment]
We should save it. Though the details of saving are still a little fuzzy. I think we should be able to add the challenge type into the saveable projects list.

**Comment by Dario-DC** [comment]
They should be able to save their code, but they should be able to reset to the seed code as well.

If it's not possible to reset the code to the seed and only to revert it to saved code, it can be a problem for labs that provide the HTML, CSS or part of the JS.

**Comment by ilenia-magoni** [comment]
I don't think it should be saved to database, it would bring back an old issue with database occupation, but it should allow for saving locally, and maybe downloading, or prompting of saving for example in github

**Comment by moT01** [comment]
I would lean towards not saving the solutions to the database as well - but am open for discussing.

The current UX is that you can save your code to local storage with ctrl+s - that isn't described anywhere however - and after you pass and submit a challenge, your saved code in local storage gets deleted.

We think we could add the "save" button to the labs so people can see that their code can be saved - and probably a warning on the completion modal that their saved code will be deleted. Maybe just a little text above the "Download my solution" button - "Your code will not be saved"

**Comment by jdwilkin4** [comment]
Now that we have moved the FSD cert to the top of the page, we are getting two type of reports about labs.

Report No1

The camper has completed a lab, moved onto the other portions of the course, wants to revisit the lab and is surprised that their code is gone.

Examples:

https://forum.freecodecamp.org/t/full-stack-developer/739994

Report No.2

The camper is currently going through a lab, needs to take a break or accidentally navigates to a different page and all of their progress is lost. As campers go through the curriculum, it is not unreasonable that they will want to take breaks and not do the more challenging labs all in one setting. 

https://forum.freecodecamp.org/t/lost-code-while-building-tribute-page/739998

We do have this message when you try to logout

<img width="1092" alt="Image" src="https://github.com/user-attachments/assets/31c28227-77f5-427a-ad6f-2c9a2d733b83" />

So maybe the wording needs to be updated to specifically mention that labs won't be saved. Because your progress is saved like for workshops and completion of quizzes and stuff.

**Comment by moT01** [comment]
So the first step here would be to add the save button to the labs - which saves to local storage. Which you can already do with ctrl+s, but the button will help. 

I think that's enough to start, but we should discuss rewording some things in the completion modal and that sign out modal after.

**Comment by mdmillerdev** [comment]
Is there potential to add a load save or upload button?

Currently, users download the file but must manually select the section for each language and copy paste it back into the respective spot in the browser, which can get annoying and could cause someone to become demotivated to continue learning if it means that they have to do extra work before they even begin.

If we added an upload button in addition tot he save button it would make the process close to as seamless as it could be using a .txt file stored anywhere. The user would still have to select the file though.

**Comment by 1Dedios** [comment]
It looks like this was changed to help wanted and no one is assigned - I'd be happy to work on this if assigned.

**Comment by majestic-owl448** [comment]
hi @1Dedios We typically do not assign issues. Instead, we accept the first pull request that comprehensively solves the issue.

Issues labeled with `help wanted` or `first timers only` are open for contributions.

Please make sure you read [our guidelines for contributing](https://contribute.freecodecamp.org/#/). We prioritize contributors following the instructions in our guide. Join us in [our chat room](https://discord.gg/PRyKn3Vbay) or [the forum](https://forum.freecodecamp.org/c/contributors/3) if you need help contributing - our community will be happy to assist you.

**Comment by 1Dedios** [comment]
@ilenia-magoni Thank you! I've thoroughly reviewed the guidelines and hope to have a contribution soon. If it's not answered in the guidelines, I'll be sure to direct my question to Discord.

**Comment by l3onhard** [comment]
Can someone point me to the folder where I would work on adding the save button to the labs?

**Comment by majestic-owl448** [comment]
@l3onhard have you checked in `client`? that's the folder where things are

**Comment by aliyasyeddd** [comment]
@ilenia-magoni the file isn't showing

**Comment by majestic-owl448** [comment]
@aliyasyeddd which file are you talking about?

**Comment by aliyasyeddd** [comment]
@ilenia-magoni folder where I can  work on adding the save button to the labs

**Comment by majestic-owl448** [comment]
@aliyasyeddd I think the folder is `client`, but could be wrong, you will need to find the file

**Comment by vivek-i** [comment]
Hey! Added the save locally button and an alert for users not signed in here: #64366

**Comment by sagi0312** [comment]
After testing, I found the data loss problem is more severe than described:

**Current behavior:** 

The code does persist to localStorage if users press Ctrl+S, but most users don't know this exists. Code is lost when:

- Changing site language
- Clicking menu items
- Any navigation except sign-out

Only sign-out has a warning - everything else silently deletes work.

**Proposed comprehensive solution:**

1. Auto-save to localStorage on every code change (like LeetCode or GreatFrontEnd) - making the existing Ctrl+S functionality automatic.
2. Show "💾 Auto-saved" indicator so users know code is safe
3. Auto-load from localStorage when returning to the lab
4. Display download button prominently in the page (for users to download the code anytime they decide to)

This builds on the existing localStorage implementation but prevents data loss for users who don't know about Ctrl+S.

**Comment by sagi0312** [comment]
Hey team! 👋

I've just opened PR https://github.com/freeCodeCamp/freeCodeCamp/pull/65137 that implements the auto-save feature with timestamp indicator and download functionality for lab challenges.

**Quick Summary:**
✅ Auto-save triggers on typing (debounced 2s), blur, unmount, and beforeunload
✅ Live timestamp indicator shows "Saved X ago" and updates every 5 seconds
✅ Download button prominently displayed in action row
✅ Built reusable `useDownloadChallenge` hook and `combineFileData` utility
✅ Refactored CompletionModal to use shared utilities (DRY principle)
✅ Timestamp resets when navigating between challenges (no stale timestamps)

The feature is fully tested locally and ready for review. Let me know if you need any clarifications or changes!

**Comment by ojeytonwilliams** [comment]
Hi @sagi0312, sorry for the delay getting back to you. 

First off, thanks for working on this and creating a solid PR. From what I can see at first glance, it works nicely and is implemented well. I'm not sure we necessarily want an auto-save feature, though. Maybe @moT01 or @jdwilkin4 could weigh in on this?

**Comment by sagi0312** [comment]
@ojeytonwilliams @moT01 @jdwilkin4 

Hi folks! Just a gentle nudge on this when you have a moment 🙂

I wanted to make sure the approach in this PR aligns with the direction you’re aiming for. It builds on the existing localStorage behavior discussed earlier in the thread and focuses on preventing accidental data loss for campers.

Happy to make any adjustments if needed - just wanted to check in. Thanks!

**Comment by sagi0312** [comment]
I opened PR #65137 to address this issue, but after 1.5 months without review or decision from maintainers, I've closed it. 

The implementation is available in the closed PR if the team wants to revisit auto-save for lab challenges in the future.

Issue remains open for someone else who has more bandwidth. Good luck!

**Comment by jdwilkin4** [comment]
It looks like this issue was opened a while ago. 
Due to the current workload of the team, I am going to close this because we don't have the bandwidth to work on this or review it at this time.

Maybe in the future, when a few things come off our plates, then we can revisit this.
